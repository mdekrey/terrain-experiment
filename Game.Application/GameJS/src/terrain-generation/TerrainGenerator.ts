import { TerrainSettings } from "./TerrainSettings";
import { initializePerlin, initializeRidgedMulti } from "../utils/LibNoiseUtils";
import { TerrainPoint } from "./TerrainPoint";
import { GameCoordinates } from "../game/GameCoordinates";

const featureOverlap = 1000;

export class TerrainGenerator {
  private readonly humidity = initializePerlin({
      lacunarity: 3.2,
      seed: 0
    });
  private readonly heat = initializePerlin({
      lacunarity: 3.2,
      seed: 1750
    });
  private readonly altitude = initializeRidgedMulti({
      lacunarity: 3.2,
      seed: 200
    });
  private readonly feature = initializeRidgedMulti({
      frequency: 1,
      lacunarity: 3.2,
      seed: 670,
      overlap: featureOverlap * 10
    });
  private readonly caveSeeds = initializePerlin({
      lacunarity: 3.2,
      seed: 900
    });
  private readonly terrainSettings: TerrainSettings;

  constructor(terrainSettings: TerrainSettings) {
    this.terrainSettings = terrainSettings;
  }

  public getTerrain(x: number, y: number): TerrainPoint {
    const altitude = this.altitude.getValue(x, y, 0);
    const heat = this.heat.getValue(x, y, 0) - Math.max(0, altitude * 2 - 1.7);
    const humidity = this.terrainSettings.humidityCurve(
      this.humidity.getValue(x, y, 0),
      heat
    );
    const feature = this.feature.getValue(x * featureOverlap, y * featureOverlap, 0);

    return new TerrainPoint(
      this.terrainSettings,
      x,
      y,
      altitude,
      heat,
      humidity,
      feature
    );
  }

  getCaveSeedAt(point: GameCoordinates) {
    return this.caveSeeds.getValue(point.x, point.y, 0) * 100000;
  }
}
