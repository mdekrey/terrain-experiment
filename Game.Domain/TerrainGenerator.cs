using System;
using System.Collections.Generic;
using System.Text;
using LibNoise;
using LibNoise.Filter;
using LibNoise.Primitive;

namespace Game.Domain
{
    public class TerrainGenerator
    {
        private static float SumFractalRange(int octaveCount)
        {
            var pow = Math.Pow(2, octaveCount);
            return (float)((pow - 1) / pow * 2);
        }
        private static readonly int OctaveCount = 6;
        private readonly float sumFractalRange = SumFractalRange(OctaveCount);
        private readonly float ridgedRange = 2.15f;
        private readonly IModule2D humidity = new SumFractal { Primitive2D = new SimplexPerlin { Seed = 0 }, OctaveCount = OctaveCount };
        private readonly IModule2D heat = new SumFractal { Primitive2D = new SimplexPerlin { Seed = 1750 }, OctaveCount = OctaveCount };
        private readonly IModule2D altitude = new RidgedMultiFractal { Primitive2D = new SimplexPerlin { Seed = 3201 }, OctaveCount = OctaveCount };
        private readonly IModule2D feature = new SumFractal { Primitive2D = new SimplexPerlin { Seed = 670 }, OctaveCount = OctaveCount, Lacunarity = 3.2f };
        private readonly IModule2D caveSeeds = new SumFractal { Primitive2D = new SimplexPerlin { Seed = 900 }, OctaveCount = OctaveCount };

        public TerrainPoint GetTerrainAt(GameCoordinate coordinates) {
            var altitude = GetAltitude(this.altitude.GetValue(coordinates));
            var heat = GetHeat(this.heat.GetValue(coordinates), altitude);
            var humidity = GetHumidity(this.humidity.GetValue(coordinates), heat);
            return new TerrainPoint(
                coordinates,
                humidity,
                heat,
                altitude,
                (FeatureBlend(feature, coordinates) / sumFractalRange).Clamp(-1, 1)
            );
        }

        private float FeatureBlend(IModule2D feature, GameCoordinate coordinates)
        {
            const float multFactor = 6000;
            const float stepSize = 100;
            // guarantees coords as 0 <= coord < stepSize
            var onex = ((coordinates.x % stepSize) + stepSize) % stepSize;
            var oney = ((coordinates.y % stepSize) + stepSize) % stepSize;
            var altx = onex + stepSize;
            var alty = oney + stepSize;
            var xWeight = onex / stepSize;
            var yWeight = oney / stepSize;
            var invXWeight = 1 - xWeight;
            var invYWeight = 1 - yWeight;

            return feature.GetValue(onex * multFactor, oney * multFactor) * xWeight * yWeight
                + feature.GetValue(altx * multFactor, oney * multFactor) * invXWeight * yWeight
                + feature.GetValue(onex * multFactor, alty * multFactor) * xWeight * invYWeight
                + feature.GetValue(altx * multFactor, alty * multFactor) * invXWeight * invYWeight;
        }

        private float GetAltitude(float inValue) => (inValue / ridgedRange).Clamp(0, 1);

        private float GetHeat(float inValue, float altitude)
        {
            var altitudeCooling = Math.Max(0, altitude * 2 - 1.7f) * 2;
            return (inValue / sumFractalRange - altitudeCooling).Clamp(-1, 1);
        }
        private float GetHumidity(float inValue, float heat)
        {
            var dehumidifyingFactor = (0.7f + heat * 0.4f);
            return (dehumidifyingFactor  * inValue / sumFractalRange).Clamp(-1, 1);
        }
    }

    static class LibNoiseExtensions
    {
        public static float GetValue(this IModule2D module, GameCoordinate coordinates) => module.GetValue(coordinates.x, coordinates.y);

        public static float Clamp(this float target, float min, float max) => Math.Min(Math.Max(target, min), max);
    }
}
