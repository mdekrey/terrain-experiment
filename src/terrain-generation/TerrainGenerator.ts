import { TerrainSettings } from "./TerrainSettings";
import { AnyDirectionGenerator, initializePerlin, initializeRidgedMulti } from "./PerlinAnyDirection";
import { clamp } from "../utils/clamp";
import { TerrainPoint } from "./TerrainPoint";

const perlinRange = Math.sqrt(3 / 4) * 2;
const clamper = clamp(0, 1 - Number.EPSILON);
const toValidRange = (v: number) => {
  return clamper(v / perlinRange + 0.5);
};

export class TerrainGenerator {
  private readonly humidity = new AnyDirectionGenerator({
    generator: initializePerlin({
      lacunarity: 3.2,
      seed: 0
    })
  });
  private readonly heat = new AnyDirectionGenerator({
    generator: initializePerlin({
      lacunarity: 3.2,
      seed: 1750
    })
  });
  private readonly altitude = new AnyDirectionGenerator({
    generator: initializeRidgedMulti({
      lacunarity: 3.2,
      seed: 3201
    })
  });
  private readonly feature = new AnyDirectionGenerator({
    generator: initializeRidgedMulti({
      lacunarity: 3.2,
      seed: 670
    }),
    overlap: 50000
  });
  private readonly terrainSettings: TerrainSettings;

  constructor(terrainSettings: TerrainSettings) {
    this.terrainSettings = terrainSettings;
  }

  public getTerrain(x: number, y: number): TerrainPoint {
    const altitude = toValidRange(this.altitude.getValue(x, y));
    const heat = toValidRange(
      this.heat.getValue(x, y) - Math.max(0, altitude * 2 - 1.7)
    );
    const humidity = this.terrainSettings.humidityCurve(
      toValidRange(this.humidity.getValue(x, y)),
      heat
    );
    const feature = toValidRange(this.feature.getValue(x * 2000, y * 2000));

    return new TerrainPoint(
      this.terrainSettings,
      x,
      y,
      altitude,
      heat,
      humidity,
      feature
    );
  }
}
