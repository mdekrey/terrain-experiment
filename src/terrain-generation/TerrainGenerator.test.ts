import { TerrainGenerator } from "./TerrainGenerator";

it("generates some terrain", () => {
    const terrain = new TerrainGenerator();

    const step = 0.01;
    const microsteps = 5;

    const noise = Array.from(Array(microsteps + 1).keys()).map(i => i / microsteps * step).map(v => terrain.getTerrain(0, v))
    console.log(noise);

});
