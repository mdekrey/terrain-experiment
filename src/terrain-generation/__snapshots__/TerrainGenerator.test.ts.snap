// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates some terrain 1`] = `
Array [
  "ColdParklands",
  "SnowyHill",
  "SnowyHill",
  "ColdParklands",
  "SnowyHill",
  "SnowyHill",
  "ConiferousForests",
  "ConiferousForests",
  "ConiferousForests",
  "ColdParklands",
  "ConiferousForests",
  "ConiferousForests",
  "ConiferousForests",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
  "ColdParklands",
]
`;

exports[`has consistent features 1`] = `
Array [
  0.794343797751154,
  0.5086951366258637,
  0.6395957121588287,
  0.868829739531859,
  0.5440129208489715,
  0.022444073871053627,
  0.8801526556778119,
  0.4489541479594801,
  0.1836638359696519,
  0.7758772819432314,
  0.3766243820746598,
  0.48306262761856306,
  0.6614045371808576,
  0.3107615648480544,
  0.3110807945055961,
  0.8164891342517362,
  0.39621496926485733,
  0.6319539654619748,
  0.8045124626795328,
  0.6922750328825203,
  0.3861697015861183,
  0.6666820818736405,
  0.6865965040396027,
  0.567648467516858,
  0.7241363502995355,
  0.6544375129016861,
  0.47629643737418986,
  0.8809240451673739,
  0.048127149428413174,
  0.6126318630234863,
]
`;
